lib.systems.examples: use four component triples for embedded

sternenseemann opened the pull request:
## Description of changes

Should solve #165836 by using triples that are interpreted the same by nixpkgs and the current gnu-config version.
Depends on/includes changes from #247325.

Tested `stdenv` build for:

* [x] riscv32
* [x] riscv64
* [x] misp64
* [x] mips
* [x] rx
* [x] arm eabi
* [x]  arm eabihf
* [x] aarch64_be
* [x] powerpc
* [x] powerpcle


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

sternenseemann committed:
lib.systems.examples: use four component triples for embedded

Thanks to a [change in gnu-config] we no longer have to rely on
gnu-config interpreting `none` as a vendor and nixpkgs as a kernel.
Instead we can specify a four component triple that corresponds to how
nixpkgs interprets the shortened one because gnu-config now accepts
them (the long form at least).

As a result, we also need to make sure that an up to date gnu-config is
used for the affected platforms (everything with kernel none).

Reference #165836.

[change in gnu-config]: https://git.savannah.gnu.org/cgit/config.git/commit/?id=998ba1414387b4ce1a519be234e1609bc7912e0c

Ericson2314 commented:
Oh yay! I had no idea whether more changes (e.g. to GCC) would be needed.

amjoseph-nixpkgs commented:
Do we really need to change nixpkgs' triples for these platforms?  Changing our triples for these platforms will be awkward for anybody already using them.

#165836 is fixed by #235230 which has the added benefit of not requiring that we always use most-canonical triples.

If we're going to require always-most-canonical triples we should do that everywhere, not just for certain platforms.

amjoseph-nixpkgs commented:
Also we need to make sure that we aren't breaking "blessed precompiled" vendor toolchains like `gcc-arm-embedded` -- we can't change their choice of triple.

This PR needs to be considered very carefully.  Please don't rush forward with this.

Ericson2314 commented:
I lean heavily towards this approach. If we discover further issues I'm happy to fix upstream packages further.

To me the basic truth is that if we don't proactively weed out the accidental complexity that is out there, no one else will. So yes we're pioneering, and there are risks, but someone has got to.

amjoseph-nixpkgs commented:
> I'm happy to fix upstream packages further.

How are you going to fix the precompiled binaries that ARM publishes?

sternenseemann commented:
> Changing our triples for these platforms will be awkward for anybody already using them.

How so? `lib.systems.examples` is intended as a way to use a platform abstract from the magic incantation required to assemble nixpkgs targeting it. If you have to hardcode the `config` anywhere, something is wrong somewhere would be my inclination.

> Also we need to make sure that we aren't breaking "blessed precompiled" vendor toolchains like `gcc-arm-embedded` -- we can't change their choice of triple.

It is still possible to pass a different `config`. Also `targetPrefix` and `${platform.config}-` don't have to be the same, so a precompiled toolchain should by and large still work (or could be easily adapted).

The way I see `lib.systems.examples` are *our* toolchains and we are within our rights to establish our own canonicality, we are likely one of the few able to do this well due to the amount of ground we cover. Taking vendor toolchains as a source of truth seems like a bad idea, since they are kind of bad at picking a sensible triple.

Ericson2314 commented:
This PR is *not* papering over things. 3 triples are ambiguous, and even if we always parsed them the exact way as GNU config, GNU config is not the only source of truth either. I would rather simply more aggressively reject ambiguous things in the triple parser --- if we don't accept something then are not biassing LLVM over GNU config or vice versa.

I made the changes to GNU config with the sense that resolved the issue --- we were out of sink with upstream, and now upstream is changed. Now discussing the vendor toolchain triples feels like moving the goal posts, to be honest. Since when have prebuilt binaries been the source of truth about anything with Nixpkgs? The explicit goal of all our work, including the  recent pure bootstrap work, is to have complete control by building the whole world from source. I don't want to give that up on the policy level.

Yes, ARM is the hardware manufacture, but I see 0 evidence that the triple they use is carefully chosen. The entire history of hardware is everyone making dubious decisions that are expedient in the moment but the introduce lots of accidental complexity. Why would the choice of triple be any different? I t seems obvious to me that some engineer at ARM simply when with the first thing that worked, and this might have predated 4 triples too.

I might be able to in fact contact an engineer who works on compilers at ARM. Would that help assuage you that we are doing nothing sketchy here?

amjoseph-nixpkgs commented:
> 3 triples are ambiguous

Of course not.  Every valid 3-triple canonicalizes to exactly one canonical triple.  `config.sub` is the totally unambiguous specification for this mapping.

> Since when have prebuilt binaries been the source of truth about anything with Nixpkgs?

You misunderstand.  We're not picking brand new triples here.  This is about not breaking a decision we've already made.  That decision happened to be made in a way that allowed people on embedded platforms to use vendor toolchains, so they've been using them.  We shouldn't break that without warning.

> I might be able to in fact contact an engineer who works on compilers at ARM. Would that help assuage you that we are doing nothing sketchy here?

Since they probably don't own a time machine, no it wouldn't.  You've seriously misunderstood my point here.  I'm not looking to them as a source of authority.  I'm pointing out that we made a decision that allowed downstream to use these toolchains, so we shouldn't break that.

I'm going to propose a compromise in my next comment.

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -124,22 +124,22 @@ rec {
 
   riscv64-embedded = {
-    config = "riscv64-none-elf";
+    config = "riscv64-unknown-none-elf";
```
```suggestion
    config = "riscv64-none-elf";
    libc = "newlib";
  };
  riscv64-embedded-canonical = {
    config = "riscv64-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -124,22 +124,22 @@ rec {
 
   riscv32-embedded = {
-    config = "riscv32-none-elf";
+    config = "riscv32-unknown-none-elf";
```
```suggestion
    config = "riscv32-none-elf";
    libc = "newlib";
  };
  riscv32-embedded-canonical = {
    config = "riscv32-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -124,22 +124,22 @@ rec {
 
   mips64-embedded = {
-    config = "mips64-none-elf";
+    config = "mips64-unknown-none-elf";
```
```suggestion
    config = "mips64-none-elf";
    libc = "newlib";
  };
  mips64-embedded-canonical = {
    config = "mips64-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -124,22 +124,22 @@ rec {
 
   mips-embedded = {
-    config = "mips-none-elf";
+    config = "mips-unknown-none-elf";
```
```suggestion
    config = "mips-none-elf";
    libc = "newlib";
  };
  mips-embedded-canonical = {
    config = "riscv64-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -153,7 +153,7 @@ rec {
 
   rx-embedded = {
-    config = "rx-none-elf";
+    config = "rx-unknown-none-elf";
```
```suggestion
    config = "rx-none-elf";
    libc = "newlib";
  };
  rx-embedded-canonical = {
    config = "rx-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -189,11 +189,11 @@ rec {
 
   arm-embedded = {
-    config = "arm-none-eabi";
+    config = "arm-unknown-none-eabi";
```
```suggestion
    config = "arm-none-elf";
    libc = "newlib";
  };
  arm-embedded-canonical = {
    config = "arm-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -189,11 +189,11 @@ rec {
   };
   armhf-embedded = {
-    config = "arm-none-eabihf";
+    config = "arm-unknown-none-eabihf";
```
```suggestion
    config = "arm-none-eabihf";
    libc = "newlib";
  };
  armhf-embedded-canonical = {
    config = "arm-unknown-none-eabihf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -204,22 +204,22 @@ rec {
 
   aarch64-embedded = {
-    config = "aarch64-none-elf";
+    config = "aarch64-unknown-none-elf";
```
```suggestion
    config = "aarch64-none-elf";
    libc = "newlib";
  };
  aarch64-embedded-canonical = {
    config = "aarch64-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -204,22 +204,22 @@ rec {
 
   aarch64be-embedded = {
-    config = "aarch64_be-none-elf";
+    config = "aarch64_be-unknown-none-elf";
```
```suggestion
    config = "aarch64_be-none-elf";
    libc = "newlib";
  };
  aarch64_be-embedded-canonical = {
    config = "aarch64_be-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -204,22 +204,22 @@ rec {
 
   ppc-embedded = {
-    config = "powerpc-none-eabi";
+    config = "powerpc-unknown-none-eabi";
```
```suggestion
    config = "powerpc-none-elf";
    libc = "newlib";
  };
  ppc-embedded-canonical = {
    config = "powerpc-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on lib/systems/examples.nix:
```diff
@@ -204,22 +204,22 @@ rec {
 
   ppcle-embedded = {
-    config = "powerpcle-none-eabi";
+    config = "powerpcle-unknown-none-eabi";
```
```suggestion
    config = "powerpcle-none-elf";
    libc = "newlib";
  };
  powerpcle-embedded-canonical = {
    config = "powerpcle-unknown-none-elf";
    libc = "newlib";
  };
```

amjoseph-nixpkgs commented on pkgs/stdenv/cross/default.nix:
```diff
@@ -80,7 +80,7 @@ in lib.init bootStages ++ [
              [ buildPackages.patchelf ]
         ++ lib.optional
-             (let f = p: !p.isx86 || builtins.elem p.libc [ "musl" "wasilibc" "relibc" ] || p.isiOS || p.isGenode;
+             (let f = p: !p.isx86 || builtins.elem p.libc [ "musl" "wasilibc" "relibc" ] || p.isiOS || p.isGenode || p.isNone;
```
```suggestion
             (let f = p: !p.isx86 || builtins.elem p.libc [ "musl" "wasilibc" "relibc" ] || p.isiOS || p.isGenode || (p.vendor=="unknown" && p.isNone);
```

amjoseph-nixpkgs reviewed (commented):
Instead of pulling the rug out from underneath anybody using these `pkgsCross.*` attributes, let's lay down a new rug and ask them to relocate themselves.

Specifically, let's split your change into two separate steps:

1. Add new `lib.examples.*` entries for the new triples you want.
2. Deprecate the old entries.
3. Remove the old entries.

I have absolutely no problem with #1; let's do it right now, today.

I'm less easily convinced of 2 and 3, but in any case since this is `lib.*`, which *is the rare part of nixpkgs that is meant to be a stable public interface*, changing things there requires `lib.warn` deprecation cycles anyways, so there's no rush.

We shouldn't pull the rug out from under people using these entries by silently changing their meaning.  If you want to migrate people to new triples, this is the way it has to be done.

Ericson2314 commented:
I am OK with the general direction of the compromise, but there is something that needs to be cleared up:

1. I called that file `examples.nix` because I intended it to be just that: mere *examples* of usage. I wanted people to write down what they wanted from scratch, using the built-in mechanisms to default missing fields to make that ergonomic.
2. `pkgsCross` was added, which became very popular, but IMO is done totally wrong: only `release.nix` should be enumerating/tabulating known-good configurations. Nixpkgs itself should only know about the current configuration. Making `pkgsCross` part of Nixpkgs such that other packages can refer to it is incorrect.

I get that in the absense of clear signals, things can become de-facto stable in ways that no one planned, but if are going to take steps because this stuff has ossified, we should clean up a few other things to reflect that:

1. Don't call a stable thing "examples"
2. *maybe* try to move this stuff out of `lib` as it is far more Nixpkgs-specific than the rest of `lib` (that was my mistake all long)
3. *maybe* deprecate `pkgsCross` and make a different thing for accessing pre-packaged package sets.

BTW in the past, I and others did change things in `example.nix` quite willy nilly, as old kludges could be removed. Maybe the era on that has closed, but I thought i should mention it.

sternenseemann committed:
arm-trusted-firmware: set HOSTCC to CC_FOR_BUILD

arm-trusted-firmware assumes that HOSTCC is `gcc` which is not
necessarily the case.

sternenseemann committed:
arm-trusted-firmware: tell Makefile abt arm-embedded's targetPrefix

Some of the firmware requires an arm-embedded toolchain to be in PATH.
The Makefile offers a similar mechanism to CROSS_COMPILE for letting it
know what naming scheme is used for the accompanying tools.

sternenseemann commented:
> `arm-trusted-firmware`

…is fixed. The problem there was arguably a packaging bug, the expression neglected to pass the compiler name along via `M0_CROSS_COMPILE` (which is not very discoverable, granted). Seems like arm is also not too strict about the name :)

> breaking change

Yes, a release notes entry doesn't hurt—I'll write one.

> poor engineering

No amount of engineering can solve a fundamentally unsolvable problem.

> Every valid 3-triple canonicalizes to exactly one canonical triple.

Firstly, this would require gnu-config as the absolute source of true, but even then it is not true, as the canonicalization changes over time. We already have to support triples that are not accepted by gnu-config and there is no telling how and if they are going to integrate them.

To my earlier point…

> Also `targetPrefix` and `${platform.config}-` don't have to be the same, so a precompiled toolchain should by and large still work (or could be easily adapted).

… I can also add that autotools actively discourages using the triple for any kind of configure tasks, recommending compilation probes instead (which are more reliable). The kind of wildcard matching they say you can do in a pinch would in most cases not catch the difference between `arm-none-eabi` and `arm-unknown-none-eabi`—because it is unreliable and ambiguous stuff to begin with. I guess it's most important role is to determine whether we are cross compiling in configure scripts (which is kind of bad for nixpkgs).

sternenseemann committed:
fusee-launcher: respect gcc-arm-embedded.targetPrefix

sternenseemann committed:
spike: don't assume naming of cross C compiler

sternenseemann commented:
Looking at `gcc-arm-embedded`, it seems to me that it is not even using cc-wrapper and is not integrated into the cross infrastructure at all.

amjoseph-nixpkgs commented:
> I am OK with the general direction of the compromise, but there is something that needs to be cleared up:
> 
>     1. I called that file `examples.nix` because I intended it to be just that: mere _examples_ of usage. I wanted people to write down what they wanted from scratch, using the built-in mechanisms to default missing fields to make that ergonomic.
> 
>     2. `pkgsCross` was added, which became very popular, but IMO is done totally wrong: only `release.nix` should be enumerating/tabulating known-good configurations. Nixpkgs itself should only know about the current configuration. Making `pkgsCross` part of Nixpkgs such that other packages can refer to it is incorrect.

**I very strongly agree with all of this.**

If we had a way to accomplish the "memoisation" that `pkgsCross` provides, I would immediately start working on a PR to delete `examples.nix`.  If somebody figures out how to memoize the cross-compiled packagesets without `pkgsCross`, let me know.

amjoseph-nixpkgs commented:
I'd like to clear the air a bit here.

I poured an insane amount of hours into #235230, which (a) fixes our triple-parser and (b) establishes a test suite so it stays fixed.  This was @sternenseeman's idea (although he originally suggested doing it by writing a *bash interpreter* in Nix, which I elected not to do).  I picked up that idea and ran with it, all the way to the finish line, at considerable personal cost.

@sternenseeman and @Ericson2314, both of you seem to have activately ignored that PR.  That's not such a big deal.  However with this PR #247490 (and #247922 and #247923) you both seem to be going to incredible lengths to Frantically Cobble weird whack-a-mole workarounds for all the bugs that *are already fixed cleanly* by #235230.  That really, really bothers me.  I can handle being ignored, and I take pride in being able to stand down when somebody proposes a better solution than mine.  But that's not what's happened here.  You two are going out of your way to piece together ugly kludges in order to avoid having to merge *or even review* #235230.

I have no idea why you are going to such great lengths to do this.  It is weird. But I'll certainly keep that in mind the next time I'm tempted to work on one of sterni's ideas.  I feel like I got bamboozled pretty badly there.

Ericson2314 commented:
@amjoseph-nixpkgs 

\[I realize we've had a bunch of pleasant conversation and collaboration more recently, but your post about from August still deserves a reply, so here it goes.\]

So the way I see is that the current parser except things which are not normal forms, just as GNU config does. Your PR adds testing, which is fantastic! But it also:

- Makes more normal forms that we do need need for expressive power (they are redundant normal forms)
- Add a *lot* complexity handling various special cases

Neither of this these things I like. I specifically want Nixpkgs to not support everything GNU config supports, but GNU config I believe supports *too many things in a chaotic way*; I don't think they would even disagree with that, but rather admit it is the case but say they have little choice because they are bound by 3 decades of back compat. Our parser is much newer, and I do not believe we bound by such decades of weight, so we should strive to come up with something simple and more elegant.

They way I would like to do that is this PR. I think we can change the configs in `exmples.nix`, I do not think think we've made any progress that they do not change. In fact, *using your tests* we should ensure going forward that the configs in `examples.nix` are always normal forms, so this sort of problem does not occur again.

-------

BTW here is one way to think about prioritizing simplicity. In issues like https://github.com/NixOS/nixpkgs/issues/242336, I've promised @trofi something  "look, stuff might seem crazy right now as we do things in weird ways, but once we've cleaned out all the major tech debt I promise it will be simpler than it was before". But if we just do https://github.com/NixOS/nixpkgs/pull/235230 alone, I am breaking that promise --- because there is no timeline for GNU config to drop its legacy cruft, and because we committing to matching a larger and more chaotic subset of it, we are permanently resigning ourselves to more complexity.

Therefore, to ensure the continued acceptance of your (and hopefully our, once I find the time :)) work overhauling how GCC is packaged and bootstrapped, which I consider *among the most foundational important efforts going on with Nixpkgs* for a variety of reasons (the interaction with the minimal bootstrap among them), I feel compelled to mind the complexity budget elsewhere. Ensuring that we don't dramatically increase the `lib.systems.parser` complexity seems to me like a good way to preserve that budget.

-------

There is one triple in `examples.nix` I am no longer trying to change, however, and that is the MinGW configs. After https://git.savannah.gnu.org/cgit/config.git/commit/?id=28ea239c53a2d5d8800c472bc2452eaa16e37af2 the decision from GNU config is loud and clear, and therefore our MinGW configs in `example.nix` should stay the same, and @amjoseph-nixpkgs commit changing the normal form to *not* be `windows-gnu` is exactly correct. I previously singled it out as the first one https://github.com/NixOS/nixpkgs/pull/235230 I wanted different, now in light of these events I am doing a 180 and saying we definitely should accept it!

Ericson2314 commented:
@sternenseemann Also in light of https://lists.gnu.org/archive/html/config-patches/2023-09/msg00048.html I think this PR should also do `vc4-elf` -> `vc4-unknown-none-elf`. We can (a) apply that patch (b) no longer prevent overriding `config.sub` for that GCC, and then everything should work along with us having one less special case.

amjoseph-nixpkgs commented:
> Our parser is much newer, and I do not believe we bound by such decades of weight, so we should strive to come up with something simple and more elegant.

I think the parser is just an implementation detail.  The taxonomy is the real question.

Are we trying to invent a new taxonomy?  I had hoped to avoid that.

Ericson2314 commented:
Invent a simpler sub-taxonomy. Everything we accept, GNU Config will accept with the same normalization (and "meaning", though that is less form). Not everything GNU Config accepts we would accept.

Ericson2314 commented:
@sternenseemann should we do this?
